File: LeetCode_Marathons/LCM2/Medium/Number_of_Islands/a.go
var globalGrid [][]byte;

func dfs(x int, y int) {
    globalGrid[x][y] = byte('0')
    n, m := len(globalGrid), len(globalGrid[0])
    var dir = [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

    for dirrIndex := 0; dirrIndex < 4; dirrIndex++ {
        nextX, nextY := x + dir[dirrIndex][0], y + dir[dirrIndex][1]
        if (nextX != -1 && nextX != n) && (nextY != -1 && nextY != m) && globalGrid[nextX][nextY] == byte('1') {
            dfs(nextX, nextY);
        }
    }
}

func numIslands(grid [][]byte) int {
    globalGrid = grid
    n, m := len(grid), len(grid[0])
    counter := 0
    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if globalGrid[i][j] == byte('1') {
                dfs(i, j)
                counter++
            }
        }
    }

    return counter
}

File: LeetCode_Marathons/LCM2/Medium/Number_of_Islands/a.java
class Solution {
    private char[][] globalGrid;

    private void dfs(int x, int y) {
        globalGrid[x][y] = '0';
        int n = globalGrid.length, m = globalGrid[0].length;
        int[][] dir = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (int dirrIndex = 0; dirrIndex < 4; dirrIndex++) {
            int nextX = x + dir[dirrIndex][0], nextY = y + dir[dirrIndex][1];
            if ((nextX != -1 && nextX != n) && (nextY != -1 && nextY != m)) {
                if (globalGrid[nextX][nextY] == '1') {
                    dfs(nextX, nextY);
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        globalGrid = grid;

        int n = grid.length, m = grid[0].length;
        int counter = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (globalGrid[i][j] == '1') {
                    dfs(i, j);
                    counter++;
                }
            }
        }

        return counter;
    }
}